# Capstone Project: Applying Machine Learning to the Crime and Dispatch datasets 

In [1]:
import os as os
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

In [2]:
from datetime import datetime

## file uploads

In [6]:
file_path_dispatch = "/Users/gimle/DATA-205-SETS/Police_Dispatched_Incidents_20240906 copy.csv"


In [8]:
file_path_crime = "/Users/gimle/DATA-205-SETS/Crime_20240906 copy.csv"


In [10]:
file_path_census = "/Users/gimle/DATA-205-SETS/DECENNIALDHC2020.P1-Data copy.csv"


In [12]:
if os.path.exists(file_path_dispatch): 
    dispatch = pd.read_csv(file_path_dispatch)
    print("dispatch created")
   
else: 
    print("youve got fix the dispatch filepath") 
    
if os.path.exists(file_path_crime): 
    crime = pd.read_csv(file_path_crime)
    print("crime created")
   
else: 
    print("youve got fix the crime filepath")
    
if os.path.exists(file_path_census): 
    census = pd.read_csv(file_path_census)
    print("crime created")
   
else: 
    print("youve got fix the census filepath") 

/var/folders/vt/8zqwv5_51777lk_0p42g3sdr0000gn/T/ipykernel_16546/2466321656.py:2: DtypeWarning: Columns (11,16) have mixed types. Specify dtype option on import or set low_memory=False.
  dispatch = pd.read_csv(file_path_dispatch)


dispatch created
crime created
crime created


# Exploratory Data analysis

In [15]:
crime.head(3)

Incident ID  Offence Code  CR Number    Dispatch Date / Time  \
0    201403032          5404  220052381  11/27/2022 11:52:08 PM   
1    201359823          5404  220000965  01/09/2022 01:18:38 AM   
2    201360056          2304  220001336  01/12/2022 02:53:25 AM   

          Start_Date_Time           End_Date_Time NIBRS Code  Victims  \
0  11/27/2022 11:52:00 PM                     NaN        90D        1   
1  01/09/2022 01:34:00 AM                     NaN        90D        1   
2  01/12/2022 02:53:00 AM  01/12/2022 03:00:00 AM        23G        1   

              Crime Name1                                  Crime Name2  ...  \
0   Crime Against Society                  Driving Under the Influence  ...   
1   Crime Against Society                  Driving Under the Influence  ...   
2  Crime Against Property  Theft of Motor Vehicle Parts or Accessories  ...   

   PRA Address Number Street Prefix Street Name Street Suffix  Street Type  \
0  594            NaN           NaN      MATENY           NaN           RD   
1  428          100.0             N   FREDERICK           NaN          AVE   
2  293         1600.0           NaN     BRADLEY           NaN          AVE   

   Latitude Longitude Police District Number             Location  
0  39.15682  -77.2665                     5D  (39.1568, -77.2665)  
1  39.09438  -77.1517                     6D  (39.0944, -77.1517)  
2  39.08126  -77.1205                     1D  (39.0813, -77.1205)  

[3 rows x 30 columns]

In [17]:
dispatch.head(3)

Incident_ID Crime Reports Crash Reports              Start Time  \
0  P1700827460     170547273           NaN  12/07/2017 11:27:39 PM   
1  P1700834957     170548471           NaN  12/14/2017 04:29:38 AM   
2  P1900298804     190043080           NaN  09/08/2019 08:55:02 PM   

                 End Time  Priority                   Initial Type  \
0  12/08/2017 04:03:37 AM         0                       BURGLARY   
1  12/14/2017 07:50:08 AM         1  DOMESTIC DISTURBANCE/VIOLENCE   
2  09/09/2019 05:13:44 AM         0          ROBBERY JUST OCCURRED   

                      Close Type                     Address           City  \
0                       BURGLARY       100 BLK  TEACHERS WAY   GAITHERSBURG   
1  DOMESTIC DISTURBANCE/VIOLENCE         12100 BLK  SIOUX PL   GAITHERSBURG   
2          ROBBERY JUST OCCURRED  13700 BLK  CONNECTICUT AVE  SILVER SPRING   

   ...    PRA CallTime CallRoute  Calltime Dispatch  Calltime Arrive  \
0  ...  419.0               78.0               99.0            121.0   
1  ...  591.0              133.0              197.0            591.0   
2  ...  337.0              147.0              218.0            351.0   

  Calltime Cleared CallRoute Dispatch Dispatch Arrive  Arrive Cleared  \
0          16557.0               20.0            21.0         16436.0   
1          12029.0               64.0           394.0         11438.0   
2          29921.0               71.0           132.0         29570.0   

          Disposition Desc                  Location  
0   ROBFIREARM-RESIDENTIAL   POINT (-77.193 39.1432)  
1          MENTALTRANSPORT  POINT (-77.2524 39.1284)  
2  LARCNYSHOPLFTNG$50-$199  POINT (-77.0808 39.0789)  

[3 rows x 26 columns]

In [19]:
crime.columns

Index(['Incident ID', 'Offence Code', 'CR Number', 'Dispatch Date / Time',
       'Start_Date_Time', 'End_Date_Time', 'NIBRS Code', 'Victims',
       'Crime Name1', 'Crime Name2', 'Crime Name3', 'Police District Name',
       'Block Address', 'City', 'State', 'Zip Code', 'Agency', 'Place',
       'Sector', 'Beat', 'PRA', 'Address Number', 'Street Prefix',
       'Street Name', 'Street Suffix', 'Street Type', 'Latitude', 'Longitude',
       'Police District Number', 'Location'],
      dtype='object')

In [21]:
dispatch.columns

Index(['Incident_ID', 'Crime Reports', 'Crash Reports', 'Start Time',
       'End Time', 'Priority', 'Initial Type', 'Close Type', 'Address', 'City',
       'State', 'Zip', 'Longitude', 'Latitude', 'Police District Number',
       'Beat', 'PRA', 'CallTime CallRoute', 'Calltime Dispatch',
       'Calltime Arrive', 'Calltime Cleared', 'CallRoute Dispatch',
       'Dispatch Arrive', 'Arrive Cleared', 'Disposition Desc', 'Location'],
      dtype='object')

In [23]:
census.columns

Index(['GEO_ID', 'NAME', 'P1_001N', 'Unnamed: 3'], dtype='object')

# Removing columns
The Beats, PRA's and Police district names can be determined from the location data, which I can fit back on in Arc GIS. Identifying the mismatches would not necessarily make for interesting results after converations with dataMontgomery
I am instead going to normalise incidents against census block data instead of either BEAT and PRA. 
Zip, state as well as other street identifiers are also not as relevant due to the existence of coordinate data. 

With regards to NIBRS codes, similar if not identical categories are represented in the crime name 2 column. 

In the case of Dispatch data the Calltimes have been removed as they will require methods I am not currently able to apply. 

In [26]:
crime.drop(['Incident ID','Block Address', 'City', 'State','Zip Code', 'Address Number', 'Street Prefix',
       'Street Name', 'Street Suffix', 'Street Type','Location','Beat', 'PRA','Sector', 'NIBRS Code', 'Police District Name'], axis=1, inplace=True) 

In [28]:
dispatch.drop(['Incident_ID','Crash Reports','Address','City', 'State', 'Zip', 'Location','Beat', 'PRA','Disposition Desc','CallTime CallRoute','Calltime Dispatch','Calltime Arrive','Calltime Arrive','Calltime Cleared','CallRoute Dispatch','Dispatch Arrive' ], axis=1, inplace=True) 

In [30]:
dispatch.isna().sum() 

Crime Reports             654675
Start Time                     0
End Time                       1
Priority                       0
Initial Type                   0
Close Type                     0
Longitude                     27
Latitude                      27
Police District Number         1
CallTime CallRoute            41
Calltime Dispatch          31638
Calltime Arrive           180215
Calltime Cleared             122
CallRoute Dispatch         31634
Dispatch Arrive           183233
Arrive Cleared            180252
dtype: int64

In [32]:
crime.isna().sum() 

Offence Code                   0
CR Number                      0
Dispatch Date / Time       46646
Start_Date_Time                0
End_Date_Time             162161
NIBRS Code                     0
Victims                        0
Crime Name1                    0
Crime Name2                    0
Crime Name3                    0
Police District Name         622
Agency                         0
Place                          0
Latitude                       0
Longitude                      0
Police District Number         0
dtype: int64

In [34]:
# Checking types
print(dispatch.dtypes)

Crime Reports              object
Start Time                 object
End Time                   object
Priority                    int64
Initial Type               object
Close Type                 object
Longitude                 float64
Latitude                  float64
Police District Number     object
CallTime CallRoute        float64
Calltime Dispatch         float64
Calltime Arrive           float64
Calltime Cleared          float64
CallRoute Dispatch        float64
Dispatch Arrive           float64
Arrive Cleared            float64
dtype: object


In [36]:
# Checking types
print(crime.dtypes)

Offence Code                int64
CR Number                   int64
Dispatch Date / Time       object
Start_Date_Time            object
End_Date_Time              object
NIBRS Code                 object
Victims                     int64
Crime Name1                object
Crime Name2                object
Crime Name3                object
Police District Name       object
Agency                     object
Place                      object
Latitude                  float64
Longitude                 float64
Police District Number     object
dtype: object


In [38]:
count_crime_1 = len(crime)
count_dispatch_1 = len(dispatch)
print(count_crime_1)
print(count_dispatch_1) 

318012
786255


In [43]:
dispatch.dropna(subset=['Longitude','Latitude',], inplace = True)
crime.dropna(subset=['End_Date_Time','Longitude','Latitude'], inplace = True)

# Fixing time and removing all entries prior to 2018
Both datasets are needlessly large and cover too much time and too many changes to internal reporting procedure. It is therefore essential to cut down the size of the dataset we are working with. 

In [46]:
# Reformatting all time and date columns
crime['Dispatch Date / Time'] = pd.to_datetime(crime['Dispatch Date / Time'], format='%m/%d/%Y %I:%M:%S %p', errors='coerce')
crime["Start_Date_Time"] = pd.to_datetime(crime['Start_Date_Time'], format='%m/%d/%Y %I:%M:%S %p', errors='coerce')
crime['End_Date_Time'] = pd.to_datetime(crime['End_Date_Time'], format='%m/%d/%Y %I:%M:%S %p', errors='coerce')

dispatch['Start Time'] = pd.to_datetime(dispatch['Start Time'], format='%m/%d/%Y %I:%M:%S %p', errors='coerce')
dispatch['End Time'] = pd.to_datetime(dispatch['End Time'], format='%m/%d/%Y %I:%M:%S %p', errors='coerce')
print("done")

done


In [48]:
# Renaming for easy of use
crime.rename(columns={"Start_Date_Time": 'Start_Time'}, inplace=True)
crime.rename(columns={'Dispatch Date / Time': 'Dispatch_Time'}, inplace=True)
crime.rename(columns={"End_Date_Time": 'End_Time'}, inplace=True)
dispatch.rename(columns={'Start Time': 'Start_Time'}, inplace=True)
dispatch.rename(columns={'End Time': 'End_Time'}, inplace=True)
dispatch.rename(columns={'Inital Type': 'Inital_Type'}, inplace=True)
dispatch.rename(columns={'End Time': 'End_Time'}, inplace=True)
print("done")

done


In [50]:
crime = crime[crime['Start_Time'] >= datetime(2018, 1, 1)]
dispatch = dispatch[dispatch['Start_Time'] >= datetime(2018, 1, 1)]
print("done")

done


In [52]:
count_crime_2 = len(crime)
count_dispatch_2 = len(dispatch)
print(count_crime_2)
print(count_dispatch_2) 

120689
637684


# Fixing location point data for easier use in ARC GIS 
This is much easier to do in while the datasets are seperate

In [55]:
crime['WKT'] = crime.apply(lambda row: f"POINT ({row['Longitude']} {row['Latitude']})", axis=1)
dispatch['WKT'] = dispatch.apply(lambda row: f"POINT ({row['Longitude']} {row['Latitude']})", axis=1)
print("done")

/var/folders/vt/8zqwv5_51777lk_0p42g3sdr0000gn/T/ipykernel_16546/2176454166.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crime['WKT'] = crime.apply(lambda row: f"POINT ({row['Longitude']} {row['Latitude']})", axis=1)


done


In [137]:
dispatch.nunique()

ID                        110202
Start_Time                635731
End_Time                  635701
Priority                       5
Initial Type                 219
Close Type                   216
Longitude                  15134
Latitude                   34367
Police District Number         9
CallTime CallRoute          1943
Calltime Dispatch          11219
Calltime Arrive            10865
Calltime Cleared           20829
CallRoute Dispatch         11037
Dispatch Arrive             7123
Arrive Cleared             17569
WKT                        64818
dtype: int64

In [103]:
print(crime['Agency'].unique())

['RCPD' 'MCPD' 'GPD' 'TPPD' 'MCSO' 'MCFM' 'OTHR']


In [105]:
print(crime['Crime Name1'].unique())

['Crime Against Property' 'Crime Against Society' 'Crime Against Person'
 'Crime Against Not a Crime' 'Crime Against Person, Property, or Society']


In [115]:
print(crime['Police District Name'].unique())

['ROCKVILLE' 'BETHESDA' nan 'SILVER SPRING' 'MONTGOMERY VILLAGE' 'WHEATON'
 'GERMANTOWN' 'TAKOMA PARK' 'OTHER']


In [133]:
set1 = set(dispatch['Close Type'])
set2 = set(dispatch['Initial Type'])

if set1 == set2:
    print("same.")
else:
    print("They're different.")
    print("only in Close Type ", set1 - set2)
    print("Only in Initial Type", set2 - set1)

They're different.
only in Close Type  {'TC - TRAFFIC COLLISION ON PATROL', 'ATT - ATTACHMENT ON PATROL', 'RAPEJA', '10 SIGNAL ALARM', 'ALARMT'}
Only in Initial Type {'SUM - SUMMONS SERVICE', 'RAPEA', 'OVERDOSE - VIA FRS - OCCURRED EARLIER', 'TRE - TRESP ON PATROL', 'WR - WRITING REPORT', 'ALARMF - FARS', 'HUNTO - HUNTING OCCURRED EARLIER - ILLEGAL - OCCURRED EARLIER', 'REFERRED TO OAS'}


In [135]:
incongruence = dispatch[dispatch['Close Type'] != dispatch['Initial Type']]
print(incongruence.shape[0])

16853


In [113]:
# Category mapping for places in the crime dataset
# If I am to one hot encode places for machine learning purposes, then they have to be simplified 

place_map = {
    
    'Cyberspace': 'Other',
    'Other/Unknown': 'Other',

    'Hotel/Motel/Etc.': 'Hospitality',
    'Restaurant': 'Hospitality',
    'Bar/Night Club': 'Hospitality',

    'Hospital/Emergency Care Center': 'Medical Facility',
    'Doctor/Dentist/Vet Office': 'Medical Facility',

    'Residence - Driveway': 'Residential Areas',
    'School - Elementary/Secondary': 'Educational Facilities',

    'Commercial - Industrial Park': 'Commercial Area',
    'Commercial - Office Building': 'Commercial Area',
    'Auto Dealership': 'Commercial Area',
    'Auto Repair': 'Commercial Area',
    'Construction Site': 'Commercial Area',
    'Industrial Site': 'Commercial Area',
    'Rental Storage Facility': 'Commercial Area',
    'Farm Facility': 'Commercial Area',
    'Dock/Wharf/Freight/Modal Terminal': 'Commercial Area',
    'Abandoned/Condemned Structures': 'Abandoned Structure',

    'Library': 'Public Space',
    'Church/Synagogue/Temple': 'Public Space',
    'Community Center': 'Public Space',
    'Recreation Center': 'Public Space',
    'Shelter-Mission/Homeless': 'Public Space',
    'Field/Open Space': 'Public Space',
    'Arena/Stadium/Fairgrounds/Coliseum': 'Public Space',
    'Golf Course': 'Public Space',
    'Lake/Waterway': 'Public Space',
    'Amusement Park': 'Public Space',
    'Pool': 'Public Space',
    'Theater': 'Public Space',

    'Government Building': 'Government Building',

    'Rest Area': 'Park or Rest Area',
    'Park': 'Park or Rest Area',

    'Parking Lot - Residential': 'Parking Lot',
    'Parking Garage - Residential': 'Parking Lot',
    'Parking Lot - Commercial': 'Parking Lot',
    'Parking Garage - Commercial': 'Parking Lot',
    'Parking Lot - Other': 'Parking Lot',
    'Parking Lot - Metro': 'Parking Lot',
    'Parking Lot - Park & Ride': 'Parking Lot',
    'Parking Garage - Metro': 'Parking Lot',
    'Parking Garage - County': 'Parking Lot',
    'Parking Garage - Other': 'Parking Lot',
    'Parking Lot - School': 'Parking Lot',
    'Parking Lot - Rec Center': 'Parking Lot',
    'Parking Lot - Church': 'Parking Lot',
    'Parking Lot - County': 'Parking Lot',

    'Street - Alley': 'Street',
    'Street - Other': 'Street',
    'Street - In vehicle': 'Street',
    'Street - Residential': 'Street',
    'Street - Commercial': 'Street',

    'Air/Bus/Train/Metro Terminal': 'Public Transit Area',
    'Street - Bus Stop': 'Public Transit Area',
    'Pedestrian Tunnel': 'Public Transit Area',

    'Grocery/Supermarket': 'Retail',
    'Gas Station': 'Retail',
    'Retail - Department/Discount Store': 'Retail',
    'Retail - Drug Store/Pharmacy': 'Retail',
    'Retail - Clothing': 'Retail',
    'Retail - Hardware': 'Retail',
    'Retail - Appliances/Electronics': 'Retail',
    'Retail - Mall': 'Retail',
    'Convenience Store': 'Retail',
    'Retail - Jewelry': 'Retail',
    'Retail - Sporting Goods': 'Retail',
    'Retail - Salon/Spa': 'Retail',
    'Retail - Beauty/Barber Shop': 'Retail',
    'Retail - Dry Cleaner': 'Retail',
    'Retail - Video Store': 'Retail',
    'Pawn Shop': 'Retail',
    'Check Cashing Est.': 'Retail',

    'Liquor Store - Beer & Wine': 'Liquor Store',
    'Liquor Store - County': 'Liquor Store',

    'School - Elementary/Secondary': 'School/Childcare',
    'School - College/University': 'School/Childcare',
    'Daycare Facility': 'School/Childcare',
    'Nursery': 'School/Childcare',

    'Residence - Driveway': 'Residence',
    'Residence - Single Family': 'Residence',
    'Residence - Apartment/Condo': 'Residence',
    'Residence - Townhouse/Duplex': 'Residence',
    'Residence - Carport': 'Residence',
    'Residence - Garage': 'Residence',
    'Residence - Nursing Home': 'Residence',
    'Residence - Shed': 'Residence',
    'Residence - Yard': 'Residence',
    'Residence - Other': 'Residence',
    'Residence - Apt Ofc/Storage': 'Residence',
    'Residence - Mobile Home': 'Residence',
}

crime['Place'] = crime['Place'].map(place_map)


In [117]:
crime.head(10)

Offence Code         ID       Dispatch_Time          Start_Time  \
2           2304  220001336 2022-01-12 02:53:25 2022-01-12 02:53:00   
3           1602  240023111 2024-05-16 11:25:24 2024-05-16 11:23:00   
6           3562  220001908 2022-01-15 22:48:27 2022-01-15 22:50:00   
15          2304  220022392 2022-05-25 17:29:23 2022-05-16 15:00:00   
17          2308  210026590 2021-07-10 20:42:35 2021-07-10 18:00:00   
18          2305  220009661 2022-03-06 12:08:40 2022-03-05 19:30:00   
19          2404  220001959 2022-01-16 12:00:09 2022-01-16 03:00:00   
26          2305  240014942 2024-03-30 15:29:13 2024-03-30 07:30:00   
29          2404  230005778 2023-02-04 17:18:17 2023-02-03 21:00:00   
30          2399  220018396 2022-04-30 15:26:00 2022-04-30 13:20:00   

              End_Time NIBRS Code  Victims             Crime Name1  \
2  2022-01-12 03:00:00        23G        1  Crime Against Property   
3  2024-05-16 14:30:00        90Z        1   Crime Against Society   
6  2022-01-15 00:00:00        35A        1   Crime Against Society   
15 2022-05-21 09:30:00        23G        1  Crime Against Property   
17 2021-07-10 18:40:00        23D        1  Crime Against Property   
18 2022-03-06 12:00:00        23F        1  Crime Against Property   
19 2022-01-16 00:00:00        240        1  Crime Against Property   
26 2024-03-30 09:30:00        23F        1  Crime Against Property   
29 2023-02-04 15:30:00        240        1  Crime Against Property   
30 2022-04-30 13:25:00        23H        1  Crime Against Property   

                                    Crime Name2  \
2   Theft of Motor Vehicle Parts or Accessories   
3                            All Other Offenses   
6                      Drug/Narcotic Violations   
15  Theft of Motor Vehicle Parts or Accessories   
17                          Theft from Building   
18                     Theft From Motor Vehicle   
19                          Motor Vehicle Theft   
26                     Theft From Motor Vehicle   
29                          Motor Vehicle Theft   
30                            All other Larceny   

                              Crime Name3 Police District Name Agency  \
2                    LARCENY - AUTO PARTS            ROCKVILLE   RCPD   
3   THREAT - TERRORISTIC - STATE OFFENSES             BETHESDA   MCPD   
6             DRUGS - MARIJUANA - POSSESS            ROCKVILLE   RCPD   
15                   LARCENY - AUTO PARTS                  NaN   MCPD   
17                    LARCENY - FROM BLDG             BETHESDA   MCPD   
18                    LARCENY - FROM AUTO             BETHESDA   MCPD   
19             AUTO THEFT - VEHICLE THEFT             BETHESDA   MCPD   
26                    LARCENY - FROM AUTO            ROCKVILLE   MCPD   
29             AUTO THEFT - VEHICLE THEFT        SILVER SPRING   MCPD   
30             LARCENY (DESCRIBE OFFENSE)   MONTGOMERY VILLAGE    GPD   

               Place  Latitude  Longitude Police District Number  \
2          Residence  39.08126   -77.1205                     1D   
3   School/Childcare   0.00000     0.0000                     2D   
6             Street  39.07219   -77.1332                     1D   
15         Residence   0.00000     0.0000                Unknown   
17            Retail  38.96184   -77.0862                     2D   
18            Street  39.02428   -77.1027                     2D   
19       Parking Lot  38.99453   -77.0363                     2D   
26            Street  39.11521   -77.1786                     1D   
29       Parking Lot  39.01519   -76.9768                     3D   
30            Retail  39.14588   -77.2199                     6D   

                          WKT  
2   POINT (-77.1205 39.08126)  
3             POINT (0.0 0.0)  
6   POINT (-77.1332 39.07219)  
15            POINT (0.0 0.0)  
17  POINT (-77.0862 38.96184)  
18  POINT (-77.1027 39.02428)  
19  POINT (-77.0363 38.99453)  
26  POINT (-77.1786 39.11521)  
29  POINT (-76.9768 39.01519)  
30  POINT (-77.2

In [72]:
dispatch.isna().sum() 

ID                             0
Start_Time                     0
End_Time                       1
Priority                       0
Initial Type                   0
Close Type                     0
Longitude                      0
Latitude                       0
Police District Number         0
CallTime CallRoute             0
Calltime Dispatch          25192
Calltime Arrive           148340
Calltime Cleared              73
CallRoute Dispatch         25194
Dispatch Arrive           151363
Arrive Cleared            148374
WKT                            0
dtype: int64

In [70]:
crime.isna().sum() 

Offence Code                 0
ID                           0
Dispatch_Time             5630
Start_Time                   0
End_Time                     0
NIBRS Code                   0
Victims                      0
Crime Name1                  0
Crime Name2                  0
Crime Name3                  0
Police District Name       272
Agency                       0
Place                        0
Latitude                     0
Longitude                    0
Police District Number       0
WKT                          0
dtype: int64

# Joining the datasets: 
Here joining the crime and dispatch datasets, keeping all
This is done by first fixing the names of the ID column which in the crime dataset is called CR number and in the dispatch dataset is called crime reports

In [62]:
crime.rename(columns={'CR Number': 'ID'}, inplace=True)
dispatch.rename(columns={'Crime Reports': 'ID'}, inplace=True)

/var/folders/vt/8zqwv5_51777lk_0p42g3sdr0000gn/T/ipykernel_16546/1865778167.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crime.rename(columns={'CR Number': 'ID'}, inplace=True)


In [64]:
dispatch['ID'] = pd.to_numeric(dispatch['ID'], errors='coerce').fillna(0).astype(int)

In [66]:
incidents = pd.merge(crime, dispatch, on='ID', how='outer', suffixes=('_crime', '_dispatch'))

In [142]:
incidents.isna().sum()

Offence Code                       557003
ID                                      0
Dispatch_Time                      569750
Start_Time_crime                   557003
End_Time_crime                     689243
NIBRS Code                         557003
Victims                            557003
Crime Name1                        557003
Crime Name2                        557003
Crime Name3                        557003
Police District Name               557599
Agency                             557003
Place                              557003
Sector                             557003
Beat_crime                         557003
PRA_crime                          557004
Latitude_crime                     557003
Longitude_crime                    557003
Police District Number_crime       557003
WKT_crime                          557003
Start_Time_dispatch                164636
End_Time_dispatch                  164637
Priority                           164636
Initial Type                      

# Matching columns and checking for errors 

The following coloumns should be synthesised 
> Police District Name - Police District Number_crime - Police District Number_dispatch ?
> WKT_dispatch - WKT_crime - distance between ?
> Start_Time_crime End_Time_crime - Start_Time_dispatch End_Time_dispatch

Remove: 
> Beat_dispatch                      164823
> PRA_dispatch                       164779
> Beat_crime                         557003
> PRA_crime                          557004
> Offence code 
> Disposition Desc                   164637
> CallTime CallRoute                 164677
> Calltime Dispatch                  190062
> Calltime Arrive                    315435
> Calltime Cleared                   164743
> CallRoute Dispatch                 190057
> Dispatch Arrive                    318475
> Arrive Cleared                     315465

In [83]:
# Yet to run
matches = incidents.dropna(subset=['ID']).shape[0]
just_dispatch = incidents[incidents['Offence Code'].isna()].shape[0]
just_crime = incidents[incidents['Priority'].isna()].shape[0]

In [ ]:
categories = ['Matches', 'Only in Crime', 'Only in Dispatch']
values = [matches, just_crime, just_dispatch]
plt.bar(categories, values)
plt.title('Comparison of Matches and Mismatches between Crime and Dispatch')
plt.ylabel('Number of Entries')
plt.show()

In [ ]:
# Save for use in dashboards
# dispatch.to_csv('/Users/gimle/DATA-205-SETS/dispatch_dash.csv', index=False)
# crime.to_csv('/Users/gimle/DATA-205-SETS/crime_dash.csv', index=False)
# incidents.to_csv('/Users/gimle/DATA-205-SETS/incidents_gis_ready.csv', index=False)